### PR TITLE
feat(hacbs_1267): adds new FBC PipelineRun bundle

### PR DESCRIPTION
**Pull request by theflockers**
Replaces https://github.com/hacbs-release/release-bundles/pull/54

Adds new Pipelines for Release FBC components.

- adds FBC Release Pipeline
- adds create-internal-request task
  
Signed-off-by: Leandro Mendes <lmendes@redhat.com>

### CODE REVIEW

File: catalog/pipeline/fbc/0.1/fbc.yaml [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+    tekton.dev/pipelines.minVersion: "0.12.1"
+    tekton.dev/tags: fbc
+spec:
+  description: >-
+    Tekton pipeline to interact with IIB service for File Based Catalogs
+  params:
+    - name: snapshot
+      type: string
+      description: The Snapshot in JSON format
+    - name: fbcServiceUrl
```
Comment (scoheb) [inline catalog/pipeline/fbc/0.1/fbc.yaml:18]: This should be IIBServiceUrl. FBC is not a service

File: catalog/pipeline/fbc/0.1/fbc.yaml [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+      type: string
+      description: FBC IIB service URL
+    - name: fbcFragment
+      type: string
+      description: FBC fragment built by HACBS
+    - name: fbcServiceSecret
```
Comment (scoheb) [inline catalog/pipeline/fbc/0.1/fbc.yaml:24]: should be IIBServiceSecret

**Review by scoheb — CHANGES_REQUESTED** [review]
various areas where we refer to FBC as a service.

File: catalog/pipeline/fbc/0.1/fbc.yaml (hunk 1 above)
Comment (theflockers) [reply catalog/pipeline/fbc/0.1/fbc.yaml]: Ok, all places referring FBC as service were changed to IIB. Please take a look again @scoheb

File: catalog/pipeline/fbc-release/0.1/README.md [hunk 3]
```diff
@@ -0,0 +1,14 @@
+# FBC Release Pipeline
+
+FBC Release Pipeline
+
+## Parameters
+
+| Name | Description | Optional | Default value |
+|------|-------------|----------|---------------|
+| snapshot | The snapshot in JSON format | No | - |
```
Comment (johnbieren) [inline catalog/pipeline/fbc-release/0.1/README.md:9]: The description in line 10 capitalizes the CRD name, so I think it should be `Snapshot` in the description (parameter name still fine)

File: catalog/pipeline/fbc-release/0.1/fbc-release.yaml [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+  annotations:
+    tekton.dev/pipelines.minVersion: "0.12.1"
+    tekton.dev/tags: release
+spec:
+  description: >-
+    Tekton release pipeline to interact with FBC Pipeline
+  params:
+    - name: snapshot
+      type: string
+      description: The snapshot in JSON format
```
Comment (johnbieren) [inline catalog/pipeline/fbc-release/0.1/fbc-release.yaml:17]: same comment as in README

File: catalog/pipeline/fbc/0.1/README.md [hunk 5]
```diff
@@ -0,0 +1,16 @@
+# FBC Pipeline
+
+FBC Pipeline
+
+## Parameters
+
+| Name | Description | Optional | Default value |
+|------|-------------|----------|---------------|
+| snapshot | The snapshot in JSON format | No | - |
```
Comment (johnbieren) [inline catalog/pipeline/fbc/0.1/README.md:9]: same comment about capitalization

File: catalog/pipeline/fbc/0.1/fbc.yaml [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+  labels:
+    app.kubernetes.io/version: "0.1"
+  annotations:
+    tekton.dev/pipelines.minVersion: "0.12.1"
+    tekton.dev/tags: fbc
+spec:
+  description: >-
+    Tekton pipeline to interact with IIB service for File Based Catalogs
+  params:
+    - name: binaryImage
```
Comment (johnbieren) [inline catalog/pipeline/fbc/0.1/fbc.yaml:15]: this doesn't align with the README

File: catalog/task/add-fbc-fragment-to-index-image/0.1/README.md [hunk 7]
```diff
@@ -0,0 +1,16 @@
+# add-fbc-fragment-to-index-image
```
Comment (johnbieren) [inline catalog/task/add-fbc-fragment-to-index-image/0.1/README.md:1]: missing samples and tests yaml for this

File: catalog/task/add-fbc-fragment-to-index-image/0.1/README.md [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+
+Calls IIB service to submit a build request to add fbc fragments to an index image
+
+## Parameters
+
+| Name | Description | Optional | Default value |
+|------|-------------|----------|---------------|
+| BINARY_IMAGE | FBC Image to be added to index image | No | |
+| IIB_SERVICE_URL | IIB Service URL | No | https://iib.engineering.redhat.com/api/v1 |
```
Comment (johnbieren) [inline catalog/task/add-fbc-fragment-to-index-image/0.1/README.md:10]: @scoheb are we okay to have this url publicly?

File: catalog/task/add-fbc-fragment-to-index-image/0.1/README.md [hunk 9]
```diff
@@ -0,0 +1,16 @@
+# add-fbc-fragment-to-index-image
+
+Calls IIB service to submit a build request to add fbc fragments to an index image
+
+## Parameters
+
+| Name | Description | Optional | Default value |
```
Comment (johnbieren) [inline catalog/task/add-fbc-fragment-to-index-image/0.1/README.md:7]: Is there a reason why all the parameters are capitalized here but not for the pipelines?

File: catalog/task/add-fbc-fragment-to-index-image/0.1/add-fbc-fragment-to-index-image.yaml [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+
+        # replaces {TOKEN} placeholder with registry credentials.
+        if [ -f /mnt/service-secret/registry_token ]; then
+            CREDS=$(cat /mnt/service-secret/registry_token)
+            sed -i "s/{TOKEN}/${CREDS}/g" ${json_input}
+        fi
+
+        # performs kerberos authentication.
+        # prioritizes using keytab over username/password.
+        #
```
Comment (johnbieren) [inline catalog/task/add-fbc-fragment-to-index-image/0.1/add-fbc-fragment-to-index-image.yaml:75]: don't need the `#` on this line

File: catalog/task/add-fbc-fragment-to-index-image/0.1/add-fbc-fragment-to-index-image.yaml [hunk 11]
```diff
@@ -0,0 +1,140 @@
...
+      default: "300"
+      description:  Timeout seconds to receive the build state
+  results:
+    - name: json_build_info
+      description: JSON build information for the requested build
+    - name: build_state
+      description: IIB Service build state
+  steps:
+    - name: add-fbc-fragment-to-index-image
+      image: quay.io/hacbs-release/release-utils:latest
```
Comment (johnbieren) [inline catalog/task/add-fbc-fragment-to-index-image/0.1/add-fbc-fragment-to-index-image.yaml:49]: I still think the change to the Dockerfile should be its own PR so we can tag that image and not have to merge this with latest

File: catalog/task/add-fbc-fragment-to-index-image/0.1/add-fbc-fragment-to-index-image.yaml [hunk 12]
```diff
@@ -0,0 +1,140 @@
...
+        /usr/bin/curl -u : --negotiate -s -X POST -H "Content-Type: application/json" -d@${json_input} --insecure \
+        "$(params.IIB_SERVICE_URL)/fbc-operations/add" | jq | tee $(results.json_build_info.path)
+
+        # checks if the previous call returned an error.
+        ! jq -e -r ".error | select( . != null )" $(results.json_build_info.path)
+      volumeMounts:
+        - name: service-secret
+          mountPath: /mnt/service-secret
+    - name: wait-for-build-state
+      image: quay.io/hacbs-release/release-utils:latest
```
Comment (johnbieren) [inline catalog/task/add-fbc-fragment-to-index-image/0.1/add-fbc-fragment-to-index-image.yaml:97]: same comment as last image line

File: catalog/task/add-fbc-fragment-to-index-image/0.1/add-fbc-fragment-to-index-image.yaml [hunk 13]
```diff
@@ -0,0 +1,140 @@
...
+      script: |
+        #!/usr/bin/env bash
+        #
```
Comment (johnbieren) [inline catalog/task/add-fbc-fragment-to-index-image/0.1/add-fbc-fragment-to-index-image.yaml:100]: don't need `#` here

File: catalog/task/add-fbc-fragment-to-index-image/0.1/add-fbc-fragment-to-index-image.yaml [hunk 14]
```diff
@@ -0,0 +1,140 @@
...
+        TASKRUN="/tmp/$$.sh"
+        cat > ${TASKRUN} <<SH
+        #!/usr/bin/env bash
+        #
```
Comment (johnbieren) [inline catalog/task/add-fbc-fragment-to-index-image/0.1/add-fbc-fragment-to-index-image.yaml:104]: don't need `#` here

File: catalog/task/add-fbc-fragment-to-index-image/0.1/add-fbc-fragment-to-index-image.yaml [hunk 15]
```diff
@@ -0,0 +1,140 @@
...
+        build_id=`jq -r ".id" $(results.json_build_info.path)`
+        state=""
+        while true; do
+            #
```
Comment (johnbieren) [inline catalog/task/add-fbc-fragment-to-index-image/0.1/add-fbc-fragment-to-index-image.yaml:108]: don't need `#` here

File: catalog/task/create-internal-request/0.1/README.md [hunk 16]
```diff
@@ -0,0 +1,13 @@
+# create-internal-request
```
Comment (johnbieren) [inline catalog/task/create-internal-request/0.1/README.md:1]: missing samples and tests dir

File: catalog/task/create-internal-request/0.1/create-internal-request.yaml [hunk 17]
```diff
@@ -0,0 +1,73 @@
...
+      description: Boolean indicating if the from_index should be overwritten      
+    - name: BUILD_TAGS      
+      type: string
+      description: List of additional tags the internal index image copy should be tagged with
+  steps:
+    - name: create-internal-request
+      image: quay.io/hacbs-release/release-utils:latest
+      script: |
+        #!/usr/bin/env bash
+        #
```
Comment (johnbieren) [inline catalog/task/create-internal-request/0.1/create-internal-request.yaml:35]: don't need `#` here

File: catalog/task/create-internal-request/0.1/README.md [hunk 18]
```diff
@@ -0,0 +1,13 @@
...
+
+Creates an InternalRequest resource to call IIB service
+
+## Parameters
+
+| Name | Description | Optional | Default value |
```
Comment (johnbieren) [inline catalog/task/create-internal-request/0.1/README.md:7]: same comment about parameter capitalization in tasks vs pipelines

File: catalog/task/create-internal-request/0.1/create-internal-request.yaml [hunk 19]
```diff
@@ -0,0 +1,73 @@
...
+      description: Index image (catalog of catalogs) the FBC fragment will be added to    
+    - name: OVERWRITE_FROM_INDEX
+      type: string
+      description: Boolean indicating if the from_index should be overwritten      
+    - name: BUILD_TAGS      
+      type: string
+      description: List of additional tags the internal index image copy should be tagged with
+  steps:
+    - name: create-internal-request
+      image: quay.io/hacbs-release/release-utils:latest
```
Comment (johnbieren) [inline catalog/task/create-internal-request/0.1/create-internal-request.yaml:32]: same comment about image

File: catalog/task/create-internal-request/0.1/create-internal-request.yaml [hunk 20]
```diff
@@ -0,0 +1,73 @@
...
+                - name: fbcFragment
+                  value: $(params.FBC_FRAGMENT)
+                - name: fromIndex
+                  value: $(params.FROM_INDEX)
+                - name: overwriteFromIndex
+                  value: $(params.OVERWRITE_FROM_INDEX)
+                - name: buildTags
+                  value: $(params.BUILD_TAGS)
+        YAML
+        oc create -f ${resourceRequest}
```
Comment (johnbieren) [inline catalog/task/create-internal-request/0.1/create-internal-request.yaml:65]: This works? I don't see us installing openshift cli in our Dockerfile

**Review by johnbieren — CHANGES_REQUESTED** [review]
Can we get more in the description of the two pipelines? I am not as involved in this as the others are, and as a user just coming to this as a user, I have no idea when I would use fbc-release pipeline and when I would use fbc pipeline. The current descriptions don't help me

File: catalog/task/create-internal-request/0.1/create-internal-request.yaml (hunk 19 above)
Comment (theflockers) [reply catalog/task/create-internal-request/0.1/create-internal-request.yaml:32]: If you are referring to your comment about the versioning in the old PR, I agree but  I would address this in a different ticket. If it is fine by you I will add the ticket to the backlog.

Comment (johnbieren) [reply catalog/task/create-internal-request/0.1/create-internal-request.yaml:32]: Can't we just split the Dockerfile change to its own PR and merge that and not have to create more tickets? If we do it the way you recommend, this gets merged with latest, then we go to change it. When we change the image, is it task version 0.2? If so, 0.1 is not guaranteed to work forever. If not, why did we make a change to the files without creating a new version?

File: catalog/task/create-internal-request/0.1/create-internal-request.yaml (hunk 20 above)
Comment (theflockers) [reply catalog/task/create-internal-request/0.1/create-internal-request.yaml]: Now should work. I added `kubectl` installation to the Dockerfile.

File: catalog/task/create-internal-request/0.1/create-internal-request.yaml (hunk 19 above)
Comment (theflockers) [reply catalog/task/create-internal-request/0.1/create-internal-request.yaml]: done.

File: catalog/pipeline/fbc-release/0.1/fbc-release.yaml [hunk 21]
```diff
@@ -0,0 +1,73 @@
...
+      description: FBC fragment built by HACBS
+    - name: fromIndex
+      type: string
+      description: Index image (catalog of catalogs) the FBC fragment will be added to
+    - name: overwriteFromIndex
+      type: string
+      description: Boolean indicating if the from_index should be overwritten
+    - name: binaryImage
+      type: string
+      description: OCP binary image to be baked into the index image.
```
Comment (johnbieren) [inline catalog/pipeline/fbc-release/0.1/fbc-release.yaml:32]: unnecessary period at end of line

**Review by johnbieren — COMMENTED** [review]
I think my only outstanding comment is asking why the task parameters are all uppercase

File: catalog/pipeline/fbc-release/0.1/fbc-release.yaml (hunk 21 above)
Comment (theflockers) [reply catalog/pipeline/fbc-release/0.1/fbc-release.yaml:32]: TBH I followed what I saw in the release.yaml for enterprise contract params. As it was a copy+paste keeping the EC part the parameters came in uppercase, hence I kept the same way for the new task. I don't have a strong reason to keep it, but after they were there thought it somehow gave a clear vision that they are parameters for the task and used in their script, and as normally shell vars are uppercase I thought they would fit properly.

I am up to change it if I am saying no sense :D @johnbieren

Comment (johnbieren) [reply catalog/pipeline/fbc-release/0.1/fbc-release.yaml:32]: I think we have been doing it `likeThisForExample` because we sometimes pass our parameters like this https://github.com/redhat-appstudio/release-service/blob/main/tekton/pipeline_run.go#L111 which wouldn't work with all uppercase. We could of course change the controller code to put the parameter all uppercase. I don't have a strong preference as long as we are consistent with our pipelines and tasks. Maybe change the parameters in this PR to be our style and we can discuss with @davidmogar once he is back from pto if we should make a ticket to change all our tekton resource parameters to uppercase as it can make it clear they are parameters and follows shell standards

File: catalog/pipeline/fbc-release/0.1/README.md [hunk 22]
```diff
@@ -0,0 +1,16 @@
...
+
+## Parameters
+
+| Name | Description | Optional | Default value |
+|------|-------------|----------|---------------|
+| snapshot | The Snapshot in JSON format | No | - |
+| enterpriseContractPolicy | JSON representation of the EnterpriseContractPolicy | No | - |
+| fbcFragment | FBC fragment built by HACBS | No | - |
+| fromIndex | Index image (catalog of catalogs) the FBC fragment will be added to | No | - |
+| overWriteFromIndex | Boolean indicating if the from_index should be overwritten | Yes | - |
```
Comment (johnbieren) [inline catalog/pipeline/fbc-release/0.1/README.md:13]: indicating if the `fromIndex` should be overwritten?

File: catalog/task/create-internal-request/0.1/create-internal-request.yaml [hunk 23]
```diff
@@ -0,0 +1,61 @@
...
+metadata:
+  name: create-internal-request
+  labels:
+    app.kubernetes.io/version: "0.1"
+  annotations:
+    tekton.dev/pipelines.minVersion: "0.12.1"
+    tekton.dev/tags: release
+spec:
+  description: >-
+      Submit a build request to add operator bundles to an index image
```
Comment (johnbieren) [inline catalog/task/create-internal-request/0.1/create-internal-request.yaml:13]: I expected this to be the same (or at least almost the same) as the description in the README for this task

File: catalog/task/create-internal-request/0.1/create-internal-request.yaml [hunk 24]
```diff
@@ -0,0 +1,61 @@
...
+      description: FBC Image to be added to index image
+    - name: fbcFragment
+      type: string
+      description: FBC fragment built by HACBS
+    - name: fromIndex
+      type: string
+      description: Index image (catalog of catalogs) the FBC fragment will be added to    
+    - name: overwriteFromIndex
+      type: string
+      description: Boolean indicating if the from_index should be overwritten
```
Comment (johnbieren) [inline catalog/task/create-internal-request/0.1/create-internal-request.yaml:29]: `s/from_index/fromIndex/`

File: catalog/pipeline/fbc-release/0.1/fbc-release.yaml [hunk 25]
```diff
@@ -0,0 +1,73 @@
...
+      description: JSON representation of the EnterpriseContractPolicy
+    - name: fbcFragment
+      type: string
+      description: FBC fragment built by HACBS
+    - name: fromIndex
+      type: string
+      description: Index image (catalog of catalogs) the FBC fragment will be added to
+    - name: overwriteFromIndex
+      type: string
+      description: Boolean indicating if the from_index should be overwritten
```
Comment (johnbieren) [inline catalog/pipeline/fbc-release/0.1/fbc-release.yaml:29]: `s/from_index/fromIndex/`

**Review by johnbieren — APPROVED** [review]

**Review by scoheb — APPROVED** [review]
